lib/job.js

```diff
@@ -27,7 +27,11 @@ class Job extends Emitter {
           client.hget(queue.toKey('jobs'), jobId, done)
         )
       )
-      .then((data) => (data ? Job.fromData(queue, jobId, data) : null));
+      .then((data) =>
+        data
+          ? Job.fromData(queue, jobId, data)
+          : queue.removeJob(jobId).then(() => null)
+      );
```

Comment:
This feels vaguely dangerous. Could we make it opt-in? What do you think?

Reply:
Sure, makes sense!

Reply:
I added the option in https://github.com/bee-queue/bee-queue/pull/296/commits/3671e4d74ad276b78d3c8a103955a8c37756e207

However, thinking about it some more this feels like a mistake, as:

 1. It isn't turned on by default so isn't as well exercised in other tests etc.
 2. If the missing job thing happens (admittedly I don't fully understand *why* it happened to us) the alternative is to have a job pass through the queue ad-infinitum. I'm not sure that's what anyone would ever want?

Just as an FYI we've been running this code in production for a little while and it hasn't caused us any problems (I'm not sure if any data has gone missing in that time, though).

Reply:
Hmm on further consideration I do want to understand how we get into this state. Maybe we reach `_finishJob` and the `multi`-`exec` doesn't remove the job from `waiting` but it's actually in there? I'm not able to focus enough to really reason about that scenario. I wonder about whether we should be adopting safety belts like this (reasonable) or whether we should avoid them and strive for end-to-end implementation correctness (pipe dream?) as a general rule for bee-queue.

Reply:
I kind of share you unease here. Maybe this whole PR is the wrong approach.

I was thinking, is it possible this comes out as a race condition between the `checkStalledJobs` script and the `_finishJob` `multi` transaction?

1. `checkStalledJobs` will move the job from `active` => `waiting` if it is stalling

2. `_finishJob` removes the job from `active` and `stalling` but not `waiting` (or `delayed` for that matter). 

I'm not quite sure what the scenario would be, I guess maybe if the machine handling the job is under heavy load it might stall, but still eventually finish the job just as another machine is moving the job to `waiting`?

In which case the solution would be simply to ensure that `_finishJob` cleans up `waiting` too.

WDYT? I'm not really sure how the issue came about but it was definitely associated with machines that were struggling and a lot of jobs stalling (see other PR).

Reply:
> In which case the solution would be simply to ensure that `_finishJob` cleans up `waiting` too.

Yeah this is where my thinking led me. I'm pretty hesitant to delete data in case there's an implementation error in `bee-queue` - we wouldn't want to outright drop data that'll never get completed. Ideally, of course, there isn't an implementation error in `bee-queue` :p

I think I'm in favor of closing the gap in `_finishJob`, though I haven't thought through the performance implications fully. Maybe that's the next step before we start deleting data? Whereas the `active` key is bounded by the number of workers, the `waiting` key is unbounded; we'd incur a worst-case `O(N)` cost for every job we complete. We can proooobably assume it'll be closer to the end of the list, and find/remove from that end - but that won't guarantee good performance. Maybe it's worth moving the `_finishJob` `MULTI` to a script so we can conditionally remove from `waiting` if it's no longer in `active`?

Reply:
> I'm pretty hesitant to delete data

Point of clarification here--the data is already deleted, it is just the references to it in the `active` (et al) lists that remain. Not sure it changes much but just wanted to be clear.

> Maybe it's worth moving the _finishJob MULTI to a script so we can conditionally remove from waiting if it's no longer in active?

I can give that a go! Might be a few days before I get round to it.

Reply:
> We can proooobably assume it'll be closer to the end of the list, and find/remove from that end - but that won't guarantee good performance.

I mean in almost all cases it won't be in there at all so that won't make much of a perf difference.

Reply:
> Point of clarification here--the data is already deleted, it is just the references to it in the `active` (et al) lists that remain. Not sure it changes much but just wanted to be clear.

Yeah good call-out - my mistake.

> I mean in almost all cases it won't be in there at all so that won't make much of a perf difference.

Using something like `LREM` will have the time complexity of `O(N)`, regardless of whether the value is in the list or not - that's the time cost I'm describing. The only way to eliminate that cost is by conditionally applying the `LREM`.

Reply:
> The only way to eliminate that cost is by conditionally applying the LREM.

Yeah, sorry I was agreeing with you but I see now my comment was a bit ambigious.